server/eclipse-project/src/main/java/us/freeandfair/corla/json/ContestJsonAdapter.java

```diff
@@ -0,0 +1,178 @@
+  
+  /**
+   * Writes a contest object.
+   * 
+   * @param the_writer The JSON writer.
+   * @param the_info The object to write.
+   */ 
+  @Override
+  public void write(final JsonWriter the_writer, 
+                    final Contest the_contest) 
+      throws IOException {
+    the_writer.beginObject();
+    the_writer.name(ID).value(the_contest.id());
+    the_writer.name(NAME).value(the_contest.name());
+    the_writer.name(COUNTY_ID).value(the_contest.county().id());
+    the_writer.name(DESCRIPTION).value(the_contest.description());
+    the_writer.name(CHOICES);
+    the_writer.beginArray();
+    for (final Choice c : the_contest.choices()) {
+      the_writer.jsonValue(Main.GSON.toJson(Persistence.unproxy(c)));
+    }
+    the_writer.endArray();
+    the_writer.name(VOTES_ALLOWED).value(the_contest.votesAllowed());
+    the_writer.name(SEQUENCE_NUMBER).value(the_contest.sequenceNumber());
+    the_writer.endObject();
+  }
+  
+  /**
+   * Reads a contest object. 
+   * 
+   * @param the_reader The JSON reader.
+   * @return the object.
+   */
+  @Override
+  public Contest read(final JsonReader the_reader) 
+      throws IOException {
+    boolean error = false;
+    Long contest_id = null;
+    String contest_name = null;
+    County county = null;
+    String description = null;
+    Choice[] choices = null;
+    Integer votes_allowed = null;
+    Integer sequence_number = null;
+    
+    the_reader.beginObject();
+    while (the_reader.hasNext()) {
+      final String name = the_reader.nextName();
+      switch (name) {
+        case ID:
+          contest_id = the_reader.nextLong();
+          break;
+        
+        case NAME:
+          contest_name = the_reader.nextString();
+          break;
+          
+        case COUNTY_ID:
+          county = Persistence.getByID(the_reader.nextLong(), County.class);
+          break;
+          
+        case DESCRIPTION:
+          description = the_reader.nextString();
+          break;
+          
+        case CHOICES:
+          choices = Main.GSON.fromJson(the_reader.nextString(), Choice[].class);
+          break;
+          
+        case VOTES_ALLOWED:
+          votes_allowed = the_reader.nextInt();
+          break;
+          
+        case SEQUENCE_NUMBER:
+          sequence_number = the_reader.nextInt();
+          break;
+          
+        default:
+          error = true;
+          break;
+      }
+    }
+    the_reader.endObject();
+    
+    // check if an identically-numbered contest exists
+    
+    if (error || contest_id == null || county == null || description == null ||
+        choices == null || votes_allowed == null) {
+      throw new JsonSyntaxException("invalid data in contest");
+    }
+    
+    if (sequence_number == null) {
+      sequence_number = 0;
+    }
+    final Contest existing = Persistence.getByID(contest_id, Contest.class);
+    final Contest result;
+    if (existing == null) {
```

Comment:
At a glance, it seems that if the user encloses a `contest_id`, but it can't be found in the DB, something has gone really wrong, and this error should maybe be unrecoverable. Why should we be permissive here and create a new `Contest`?

Reply:
Nobody should ever be sending us contests to unserialize - there's no endpoint that accepts them as input. But if they do, there's no good reason why we shouldn't unserialize them, and let any error that might occur when they later want to save the thing to the database occur later... You could imagine, for example, an endpoint that actually lets you specify contest details, in which case you would in fact be creating new objects.

I will, however, readily admit that for _some_ classes - mostly those that we only create internally and that there is no legitimate reason to ever have created externally - I have not been nearly as forgiving in the unserialization phase.